src/providers/java.rs

```diff
@@ -29,15 +31,17 @@ impl Provider for JavaProvider {
     }
 
     fn get_build_plan(&self, app: &App, _env: &Environment) -> Result<Option<BuildPlan>> {
-        let mut setup = Phase::setup(Some(vec![Pkg::new("jdk")]));
-
+        let mut setup: Phase;
         let build = if self.is_using_gradle(app) {
-            setup.add_nix_pkgs(vec![Pkg::new("gradle")]);
+            let pkgs = self.get_jdk_and_gradle_pkgs(app)?;
+            setup = Phase::setup(Some(pkgs));
+
             let gradle_exe = self.get_gradle_exe(app);
             let mut build = Phase::build(Some(format!("{} build", gradle_exe)));
-            build.add_cache_directory(".gradle");
+            build.add_cache_directory("~/.gradle");
             build
         } else {
+            setup = Phase::setup(Some(vec![Pkg::new("jdk")]));
```

Comment:
Unrelated to gradle, but we might also want to check for a specific version of the `jdk` (https://github.com/railwayapp/nixpacks/issues/455)

I'm only mentioning since you are in Java land already :)

Reply:
Awesome!
I'm still in the Java context, let me shoot another PR for this to decrease the noise here!